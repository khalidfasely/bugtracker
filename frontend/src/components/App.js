function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
