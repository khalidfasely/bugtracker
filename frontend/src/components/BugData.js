const BugData = ({ bug }) => (
    <div>
        <p>{bug?.id}</p>
        <p>{bug?.user?.username}</p>
        <p>{bug?.title}</p>
        <p>{bug?.description}</p>
        <p>{bug?.time}</p>
        <p>{bug?.classification}</p>
        <p>{`${bug?.active}`}</p>
        <div>
            {bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <br />
        <div>
            {bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;