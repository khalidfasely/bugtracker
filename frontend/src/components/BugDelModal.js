import Modal from "react-modal";

const BugDelModal = ({ delModalOpen, setDelModalOpen, deleteBug }) => (
    <Modal
        isOpen={delModalOpen}
        onRequestClose={() => setDelModalOpen(false)}
    >
        If you delete this bug there is no way to return it!
        <button onClick={deleteBug} data-testid='delete_button'>Delete</button>
    </Modal>
);

export default BugDelModal;