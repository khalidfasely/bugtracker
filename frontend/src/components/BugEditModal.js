import Modal from "react-modal";

import NewBug from "./NewBug";

const BugEditModal = ({
    editModalOpen, setEditModalOpen,
    users, bug
}) => (
    <Modal
        isOpen={editModalOpen}
        onRequestClose={() => setEditModalOpen(false)}
    >
        <NewBug users={users} isEdit={true} bug={bug} setEditModalOpen={setEditModalOpen} />
    </Modal>
);

export default BugEditModal;