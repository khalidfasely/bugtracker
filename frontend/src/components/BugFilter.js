import { connect } from "react-redux";
import { setTextFilter, sortByDefault, sortByPriority } from "../actions/bugFilter";

export const BugFilter = ({ filters, setTextFilter, sortByDefault, sortByPriority }) => {
    const onSortByChange = (e) => {
        if (e.target.value === 'priority') {
            sortByPriority();
            return;
        };
        sortByDefault();
    };

    return (
        <div className="bugs-filter">
            <div className="input-group__item">
                <input
                    data-testid='search_input'
                    type="text"
                    className="text-input"
                    placeholder="Search bugs"
                    value={filters.text}
                    maxLength={35}
                    onChange={(e) => setTextFilter(e.target.value)}
                />
            </div>
            <div className="input-group__item">
                <select
                    data-testid='select_input'
                    className="select"
                    value={filters.sortBy}
                    onChange={onSortByChange}
                >
                    <option value="">Date</option>
                    <option value="priority">Priority</option>
                </select>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.bugFilter
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByPriority: () => dispatch(sortByPriority()),
    sortByDefault: () => dispatch(sortByDefault())
});

export default connect(mapStateToProps, mapDispatchToProps)(BugFilter);