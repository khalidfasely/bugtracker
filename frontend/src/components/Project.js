import { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { connect } from 'react-redux';

import { startSetProject } from "../actions/project";
import ProjectItem from "./ProjectItem";

export const Project = ({ startSetProject }) => {
    const params = useParams();
    //const [project, setProject] = useState({name: "default", time: "default", id: "default", user: "default"});
    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');
    //if (!projects.users_with.include(uname)) {
    //    return <div>You're not able to see this project</div>
    //}

    useEffect(() => {
        startSetProject(params.pid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };

            //setProject(projectItem);
            //Instead of rendering the result.project
            // I do set the project on redux then render it from the projectItem
            // otherwise if I want to render it here it render the project before on the redux
            // With that saying I render the project info AFTER the project stored in redux
        })
        .then(() => setAbleToSee(true));
    }, [params])

    if (notFound) {
        return <div>Project Not Found! -- {notFound}</div>;
    }

    return (
        <div>
            {
                ableToSee ?
                <div><ProjectItem /></div> :
                <div>You're not able to see this project</div>
            }
        </div>
    );
};

//const mapStateToProps = (state) => ({
//    // Not uses yet
//    projectItem: state.projectItem.project
//});

const mapDispatchToProps = (dispatch) => ({
    startSetProject: (pid) => dispatch(startSetProject(pid))
});

export default connect(undefined, mapDispatchToProps)(Project);