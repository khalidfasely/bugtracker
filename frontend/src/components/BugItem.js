import { useState, useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { connect } from "react-redux";

import Comment from "./Comment";
import BugData from "./BugData";
import NewComment from "./NewComment";
import { startSetDeleteBug } from "../actions/project";
import { startSetSelectUsers } from '../actions/selectUsers';
import BugDelModal from "./BugDelModal";
import BugEditModal from "./BugEditModal";

export const BugItem = ({
        uname, bug, comments, startSetDeleteBug,
        users, startSetSelectUsers
    }) => {
    const history = useNavigate();

    const [delModalOpen, setDelModalOpen] = useState(false);
    const [editModalOpen, setEditModalOpen] = useState(false);

    // Set select users
    useEffect(() => startSetSelectUsers(bug.on_project), []);

    const deleteBug = () => {
        startSetDeleteBug(bug.id).then(() => {
            history(`/project/${bug.on_project}`);
            setDelModalOpen(false);
        });
    };

    return (
        <>
            {
                bug?.users_with?.includes(uname) ?
                <div>
                    <BugData bug={bug} />
                    <hr />
                    {
                        uname===bug.user.username &&
                        <div>
                            <button onClick={() => setDelModalOpen(true)}>Delete</button>
                            <button onClick={() => setEditModalOpen(true)}>Edit</button>
                        </div>
                    }
                    <hr />
                    <NewComment bugId={bug.id} />
                    <hr />
                    {comments?.map(comment => <div key={comment.id}><Comment comment={comment} /></div>)}
                </div> :
                <div>Not able to see this bug! <Link to={`/project/${bug.on_project}`}>Create one!</Link></div>
            }
            <BugDelModal
                delModalOpen={delModalOpen}
                setDelModalOpen={setDelModalOpen}
                deleteBug={deleteBug}
            />
            <BugEditModal
                editModalOpen={editModalOpen}
                setEditModalOpen={setEditModalOpen}
                users={users}
                bug={bug}
            />
        </>
    );
}

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    bug: state.bugItem.bug,
    comments: state.bugItem.comments,
    users: state.selectUsers
});

const mapDispatchToProps = (dispatch) => ({
    startSetDeleteBug: (bid) => dispatch(startSetDeleteBug(bid)),
    startSetSelectUsers: (pid) => dispatch(startSetSelectUsers(pid))
});

export default connect(mapStateToProps, mapDispatchToProps)(BugItem);