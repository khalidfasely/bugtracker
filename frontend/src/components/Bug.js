import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);