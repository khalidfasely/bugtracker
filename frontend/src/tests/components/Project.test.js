import React from "react";
import { render, act } from '@testing-library/react';

import { Project } from "../../components/Project";

jest.mock('../../components/ProjectItem', () => () => <div>ProjectItem</div>);

test('Should render Project component correctly', async () => {
    const promise = Promise.resolve({});
    const props = {
        // mock a async function
        startSetProject: jest.fn(() => promise)
    }
    const { asFragment } = render(<Project {...props} />);
    expect(asFragment()).toMatchSnapshot();
    await act(() => promise);
});

test('Should render Project component correctly after waiting the project to come from backend', async () => {
    const promise = Promise.resolve({});
    const props = {
        startSetProject: jest.fn(() => promise)
    }
    const { asFragment } = render(<Project {...props} />);
    await act(() => promise);
    expect(asFragment()).toMatchSnapshot();
});