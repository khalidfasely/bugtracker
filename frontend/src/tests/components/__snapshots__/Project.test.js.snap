// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Project component correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      You're not able to see this project
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render Project component correctly after waiting the project to come from backend 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        ProjectItem
      </div>
    </div>
  </div>
</DocumentFragment>
`;
